lib/spack/spack/directives.py

```diff
@@ -621,36 +641,40 @@ def _execute_provides(pkg):
             if pkg.name == provided_spec.name:
                 raise CircularReferenceError("Package '%s' cannot provide itself." % pkg.name)
 
-            if provided_spec not in pkg.provided:
-                pkg.provided[provided_spec] = set()
-            pkg.provided[provided_spec].add(when_spec)
+            provided_set = pkg.provided.setdefault(when_spec, set())
+            provided_set.add(provided_spec)
```

Comment:
Is there any technical reason for this :point_up:

Reply:
I guess my question would be why not a `set`?  The virtuals provided need to be unique... why use a list and list the same thing twice?

Reply:
> why use a list and list the same thing twice?

I'm not asking that, see "(eventually constrained if multiple directives refer to the same virtual under the same condition)".

What I'm wondering is why storing:
```python
provides("mpi", when="+foo")
provides("mpi@2:", when="+foo")
provides("blas", when="+foo")
```
as `{"mpi", "mpi@2:", "blas"}` rather than just `["mpi@2:", "blas"]`.

Reply:
This is similar to what Harmen has been asking. `provides` is not the same as `depends_on` — with dependencies, you need to intersect all the constraints on each dep. With `provides()` you need to union them.

In the example above, you’re saying the thing provides `mpi` whatever version when `+foo`, and `mpi` 2 or higher as well. The union of those is `mpi`, so we *could* add a union operation and use it here; we just don’t happen to have one.

most packages don’t write their provides this way — you typically want something like what we use for MPI:

```
provides(“mpi@:1”, when=“@1:”)
provides(“mpi@:2”, when=“@3:”)
```

here, version 2 introduces mpi 1 support and version 3 introduces mpi 2 support (but keeps mpi 1 support as well — the provides are unioned where the whens overlap).

Providers are producers (providing all features/versions declared) not consumers (for whom you must satisfy all requirements declared); they’re on the other side of the depends_on directive here so it has to be flipped.

Reply:
> Lower (right-open) bounds don’t really make sense for provides() — that would mean you provide every future version of something… which you can’t possibly do because future versions aren’t released.

This is slightly OT, but I agree with that. This is also inline with what I was proposing yesterday at the meeting: enforce virtual packages to be defined and have a list of available versions. Once we have that, we can also have a proper audit message for these directives, if they ever omit an upper bound.